### PR TITLE
router plugin: interceptor with xDS router

### LINKED ISSUE #1604: Support obtaining Istio routing and load balancing configurations based on the xDS protocol
# Value
Sermant is a Proxyless Service Mesh Solution with better performance.The support for the xDS protocol enables Sermant to better integrate into the Cloud Native Ecosystem. In Istio, Sermant can replace Envoy as the data plane to obtain directly the configuration of the control plane, and avoid performance loss and architecture complexity caused by Sidecar containers.
# Objective
Sermant can connect to the control plane of Istio through the xDS API interface to obtain the configurations of CDS, EDS, LDS, and RDS, and parse configurations to data model of the Sermant. Sermant framework and plugins will  implement gradually capabilities such as service discovery、routing configuration and load balancing configuration, and the the host microservice can implements the Proxyless Mesh capability by mounting Sermant to  invoke directly the upstream service registered with the control plane, such as istiod.
# RoadMap
**v2.0.0 (2024/06/30)**
Support xDS protocol for service discovery in Sermant framework
**v2.1.0 (2024/09/30)**
Support xDS protocol for routing configuration and load balancing configuration
**v2.2.0 (2024/12/30)**
Support xDS protocol for flow control configuration

### PR DESCRIPTION
**Pull request by daizhenyu**
**What type of PR is this?**

Feature.

**What this PR does / why we need it?**

router plugin: interceptor with xDS router, include:
1. HttpClient4.x
2. OkHttp2.x+
3. HttpUrlConnection
4. SpringCloud

**Which issue(s) this PR fixes？**

Fixes #1604

**Does this PR introduce a user-facing change?**

No

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+ * @since 2024-09-06
+ */
+public class OkHttpClientInterceptorChainDeclarer extends BaseRegistryPluginAdaptationDeclarer {
+    private static final String[] ENHANCE_CLASSES = {
+            "com.squareup.okhttp.Call$ApplicationInterceptorChain"
+    };
+
+    @Override
+    public ClassMatcher getClassMatcher() {
+        return ClassMatcher.nameContains(ENHANCE_CLASSES);
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java]: use nameequals for one element

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    @Override
+    public InterceptDeclarer[] getInterceptDeclarers(ClassLoader classLoader) {
+        return new InterceptDeclarer[]{
+                InterceptDeclarer.build(MethodMatcher.nameContains("proceed"),
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java]: use nameequals for one element

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java [hunk 3]
```diff
@@ -50,28 +57,47 @@ public class BaseLoadBalancerInterceptor extends AbstractInterceptor {
...
+        if (routerConfig.isEnabledXdsRoute()) {
+            serviceInstanceByXdsRoute = XdsRouterHandler.INSTANCE
+                    .getServiceInstanceByXdsRoute(name, requestData.getPath(),
+                            BaseHttpRouterUtils.processHeaders(requestData.getTag()));
+        }
+        if (serviceInstanceByXdsRoute != null && !serviceInstanceByXdsRoute.isEmpty()) {
+            context.skip(Collections.unmodifiableList(SpringRouterUtils
+                    .getSpringCloudServiceInstanceByXds(serviceInstanceByXdsRoute)));
+            return context;
         }
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: xds logic can be extracted as a function

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java [hunk 4]
```diff
@@ -98,4 +112,32 @@ public ExecuteContext onThrow(ExecuteContext context) {
...
+
+    private Map<String, String> getHeaders(HttpRequestBase httpRequest) {
+        Map<String, String> headerMap = new HashMap<>();
+        for (Header header : httpRequest.getAllHeaders()) {
+            headerMap.putIfAbsent(header.getName(), header.getValue());
+        }
+        return headerMap;
+    }
+
+    private void handleXdsRouterAndUpdateHttpRequest(Object[] arguments, HttpRequestBase httpRequest) {
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: input arguments[0] is enough

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java [hunk 5]
```diff
@@ -40,6 +48,10 @@
  * @since 2022-10-25
  */
 public class HttpClient4xInterceptor extends MarkInterceptor {
+    private static final Logger LOGGER = LoggerFactory.getLogger();
+
+    private RouterConfig routerConfig = PluginConfigManager.getPluginConfig(RouterConfig.class);
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java:53]: RouterConfig can be defined in MarkInterceptor

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/LoadBalancerInterceptor.java [hunk 6]
```diff
@@ -58,6 +68,20 @@ public ExecuteContext before(ExecuteContext context) {
...
+        if (routerConfig.isEnabledXdsRoute()) {
+            serviceInstanceByXdsRoute = XdsRouterHandler.INSTANCE
+                    .getServiceInstanceByXdsRoute(serviceId, requestData.getPath(),
+                            BaseHttpRouterUtils.processHeaders(requestData.getTag()));
+        }
+        if (serviceInstanceByXdsRoute != null && !serviceInstanceByXdsRoute.isEmpty()) {
+            arguments[0] = SpringRouterUtils
+                    .getSpringCloudServiceInstanceByXds(serviceInstanceByXdsRoute);
+            return context;
+        }
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/LoadBalancerInterceptor.java]: extract these lines

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/OkHttp3ClientInterceptor.java [hunk 7]
```diff
@@ -93,4 +107,36 @@ private Optional<Request> getRequest(ExecuteContext context) {
...
+                .build();
+    }
+
+    private void handleXdsRouterAndUpdateHttpRequest(Object obj, Request request) {
+        Headers headers = request.headers();
+        URI uri = request.url().uri();
+        String host = uri.getHost();
+
+        // use xds route to find a service instance, and modify url by it
+        if (routerConfig.isEnabledXdsRoute() && BaseHttpRouterUtils.isXdsRouteRequired(host)) {
```
Comment (lilai23) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/OkHttp3ClientInterceptor.java]: I think ‘routerConfig.isEnabledXdsRoute()’ can be evaluated at first so you do not have to create previous objects.  Modify other places

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java [hunk 8]
```diff
@@ -50,28 +57,47 @@ public class BaseLoadBalancerInterceptor extends AbstractInterceptor {
...
+        BaseLoadBalancer loadBalancer = (BaseLoadBalancer) object;
+        String name = loadBalancer.getName();
+        RequestData requestData = getRequestData().orElse(null);
+
+        // use xds route to find service instances
+        Set<ServiceInstance> serviceInstanceByXdsRoute = null;
+        if (routerConfig.isEnabledXdsRoute()) {
+            serviceInstanceByXdsRoute = XdsRouterHandler.INSTANCE
+                    .getServiceInstanceByXdsRoute(name, requestData.getPath(),
+                            BaseHttpRouterUtils.processHeaders(requestData.getTag()));
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: requestData may be null

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java [hunk 9]
```diff
@@ -50,28 +57,47 @@ public class BaseLoadBalancerInterceptor extends AbstractInterceptor {
...
+        }
+        if (serviceInstanceByXdsRoute != null && !serviceInstanceByXdsRoute.isEmpty()) {
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: CollectionUtils is recommended.

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java [hunk 10]
```diff
@@ -98,4 +112,32 @@ public ExecuteContext onThrow(ExecuteContext context) {
...
+                ServiceInstance instance = serviceInstanceOptional.get();
+                try {
+                    httpRequest.setURI(new URI(BaseHttpRouterUtils.rebuildUrlByXdsServiceInstance(uri, instance)));
+                    arguments[0] = new HttpHost(instance.getHost(), instance.getPort());
+                } catch (URISyntaxException e) {
+                    LOGGER.log(Level.WARNING, "Create uri using xds service instance failed.", e.getMessage());
+                }
+            }
+        }
+    }
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: It is recommended that the unqualified items be directly returned to reduce hierarchical nesting.

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java [hunk 11]
```diff
@@ -50,31 +62,33 @@ public class HttpClient4xInterceptor extends MarkInterceptor {
...
+                String headerValue = header.getValue();
+                Map<String, List<String>> stringListMap = FlowContextUtils.decodeTags(headerValue);
+                flowTags.putAll(stringListMap);
             }
         }
+        if (flowTags.size() > 0) {
+            ThreadLocalUtils.setRequestData(new RequestData(
+                    flowTags, uri.getPath(), httpRequest.getRequestLine().getMethod()));
+        }
         return context;
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: It is recommended that you directly return context when headers is empty or null.

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/SpringRouterUtils.java [hunk 12]
```diff
@@ -53,21 +75,29 @@ public static Map<String, String> getMetadata(Object obj) {
...
         if (!CollectionUtils.isEmpty(parameters)) {
             // The request header is changed to lowercase in the HTTP request
             parameters.forEach((key, value) -> metadata.putIfAbsent(key.toLowerCase(Locale.ROOT), value));
         }
         AppCache.INSTANCE.setMetadata(metadata);
     }
+
+    private static org.springframework.cloud.client.ServiceInstance convertServiceInstance(
+            io.sermant.core.service.xds.entity.ServiceInstance xdsServiceInstance) {
+        String instanceId = xdsServiceInstance.getHost() + xdsServiceInstance.getPort();
```
Comment (hanbingleixue) [inline sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/SpringRouterUtils.java]: You are advised to add a connector.

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java (hunk 1 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java (hunk 2 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/declarer/OkHttpClientInterceptorChainDeclarer.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java (hunk 4 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/OkHttp3ClientInterceptor.java (hunk 7 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/OkHttp3ClientInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java (hunk 8 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java (hunk 9 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java (hunk 10 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: use guard clauses

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/LoadBalancerInterceptor.java (hunk 6 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/LoadBalancerInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java (hunk 3 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/BaseLoadBalancerInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java (hunk 11 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/interceptor/HttpClient4xInterceptor.java]: done

File: sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/SpringRouterUtils.java (hunk 12 above)
Comment (daizhenyu) [reply sermant-plugins/sermant-router/spring-router-plugin/src/main/java/io/sermant/router/spring/utils/SpringRouterUtils.java]: done

**Review by hanbingleixue — APPROVED** [review]

**Review by lilai23 — APPROVED** [review]

**Pull request merged**